### PR TITLE
Fix: ignore null literals for non nullable types

### PR DESCRIPTION
**Pull request by XinyiQiao**
Hi team, could you please review this draft PR on issue [#1701](https://github.com/questdb/questdb/issues/1701)?

We fixed this issue by always return:
false for col = null when col's type is [char, byte, bool, short]
true for col != null when col's type is [char, byte, bool, short]

For the test` io.questdb.griffin.InsertNullTest#testInsertNull`, we had an issue with \n in query result and our current fix was to remove all \n characters in TestUtils.assertEquals. Not sure if this follows the best practice and would appreciate any insight!

### CODE REVIEW

**Comment by marregui** [comment]
Thank you very much for the contribution!. The approach is correct. 

Looking into the subject, it turns out that type `char` needs in fact to be nullable. At present it is not documented as such (I will update the documentation), however the argument is that a char is a one character string, and the latter is nullable. In both cases the empty value `''` and NULL should be the same. Type `char` is nullable. 

These tests would need to pass:

```
create table tab_char(v char);
insert into tab_char select rnd_char() from long_sequence(9);
insert into tab_char values(NULL);
tab_char where v is NULL --> 1 row ;
tab_char where v is NOT NULL --> 9 rows ;
select cast(0 AS CHAR) IS NULL --> true;
select cast(0L AS CHAR) IS NULL --> true;
select cast('' AS CHAR) IS NULL --> true;
select cast(NULL AS CHAR) IS NULL --> true;
select cast(false AS CHAR) IS NULL --> false;
select cast(0 AS CHAR) IS NOT NULL --> false;
select cast(0L AS CHAR) IS NOT NULL --> false;
select cast('' AS CHAR) IS NOT NULL --> false;
select cast(NULL AS CHAR) IS NOT NULL --> false;
select cast(false AS CHAR) IS NOT NULL --> true;
```

You could have a look at the above, or simply revert/ignore `char` related code in this PR, to make it about the only non nullable types `short`, `byte` and `boolean`.

For `short` these tests would need to pass:

```
create table tab_short(v short);
insert into tab_short select rnd_short() from long_sequence(9);
insert into tab_short values(NULL);
tab_short where v is NULL --> empty ;
tab_short where v is NOT NULL --> all 10 rows ;
select cast(0 AS SHORT) IS NULL --> false;
select cast(0L AS SHORT) IS NULL --> false;
select cast('' AS SHORT) IS NULL --> false;
select cast(NULL AS SHORT) IS NULL --> false;
select cast(false AS SHORT) IS NULL --> false;
select cast(0 AS SHORT) IS NOT NULL --> true;
select cast(0L AS SHORT) IS NOT NULL --> true;
select cast('' AS SHORT) IS NOT NULL --> true;
select cast(NULL AS SHORT) IS NOT NULL --> true;
select cast(false AS SHORT) IS NOT NULL --> true;
```

For `byte` these tests would need to pass:

```
create table tab_byte(v byte);
insert into tab_byte select rnd_byte() from long_sequence(9);
insert into tab_byte values(NULL);
tab_byte where v is NULL --> empty ;
tab_byte where v is NOT NULL --> all 10 rows ;
select cast(0 AS BYTE) IS NULL --> false;
select cast(0L AS BYTE) IS NULL --> false;
select cast('' AS BYTE) IS NULL --> false;
select cast(NULL AS BYTE) IS NULL --> false;
select cast(false AS BYTE) IS NULL --> false;
select cast(0 AS BYTE) IS NOT NULL --> true;
select cast(0L AS BYTE) IS NOT NULL --> true;
select cast('' AS BYTE) IS NOT NULL --> true;
select cast(NULL AS BYTE) IS NOT NULL --> true;
select cast(false AS BYTE) IS NOT NULL --> true;
```

For `boolean` these tests would need to pass, however there are a couple that hit the same bug. You can annotate tests to be ignored with `@Ignore` (please also add a `TODO` explaining that it evidences a bug). Alternatively you can go down the rabbit hole and get a closer look at the bug, which may be simple to fix in this PR.

```
create table tab_boolean(v boolean);
insert into tab_boolean select rnd_boolean() from long_sequence(9);
insert into tab_boolean values(NULL);
tab_boolean where v is NULL --> empty ;
tab_boolean where v is NOT NULL --> all 10 rows ;
select cast(0 AS BOOLEAN) IS NULL --> false;
select cast(0L AS BOOLEAN) IS NULL --> false;
select cast('' AS BOOLEAN) IS NULL --> false;              <---- HERE
select cast(NULL AS BOOLEAN) IS NULL --> false;
select cast(false AS BOOLEAN) IS NULL --> false;
select cast(0 AS BOOLEAN) IS NOT NULL --> true;
select cast(0L AS BOOLEAN) IS NOT NULL --> true;
select cast('' AS BOOLEAN) IS NOT NULL --> true;       <---- HERE
select cast(NULL AS BOOLEAN) IS NOT NULL --> true;
select cast(false AS BOOLEAN) IS NOT NULL --> true;
```

Thank you again @XinyiQiao , would you be able to update the PR.

As per the CLA issue, hmm, would you be able to update your fork and then make these five commits a single commit authored by you only?, I suppose you would then need to do a push force.

**Comment by XinyiQiao** [comment]
Hi Miguel, thank you so much for your feedback! I've reverted char related code and working on the tests now!

**Comment by XinyiQiao** [comment]
Hi Miguel @marregui I've added the tests you mentioned and reverted the change in TestUtils. Could you please have a look at the updated PR? Thank you!

File: core/src/main/java/io/questdb/griffin/engine/functions/eq/EqBooleanFunctionFactory.java [hunk 1]
```diff
@@ -61,6 +61,9 @@ public Func(Function left, Function right) {
 
         @Override
         public boolean getBool(Record rec) {
+            if(left.isNullConstant() || right.isNullConstant()) {
+                return negated || false;
```
Comment (marregui) [inline core/src/main/java/io/questdb/griffin/engine/functions/eq/EqBooleanFunctionFactory.java:65]: Table of Truth:

| negated | result |
|---------|-------|
| false      | false   |
| true       | true     |

the `|| false` is thus redundant.

File: core/src/main/java/io/questdb/griffin/engine/functions/eq/EqByteFunctionFactory.java [hunk 2]
```diff
@@ -61,6 +61,9 @@ public Func(Function left, Function right) {
 
         @Override
         public boolean getBool(Record rec) {
+            if(left.isNullConstant() || right.isNullConstant()) {
+                return negated || false;
```
Comment (marregui) [inline core/src/main/java/io/questdb/griffin/engine/functions/eq/EqByteFunctionFactory.java:65]: ditto 

Table of Truth:

| negated | result |
|---------|-------|
| false      | false   |
| true       | true     |

the `|| false` is thus redundant.

File: core/src/main/java/io/questdb/griffin/engine/functions/eq/EqShortFunctionFactory.java [hunk 3]
```diff
@@ -61,6 +61,9 @@ public Func(Function left, Function right) {
 
         @Override
         public boolean getBool(Record rec) {
+            if (left.isNullConstant() || right.isNullConstant()) {
+                return negated || false;
```
Comment (marregui) [inline core/src/main/java/io/questdb/griffin/engine/functions/eq/EqShortFunctionFactory.java:65]: ditto 

Table of Truth:

| negated | result |
|---------|-------|
| false      | false   |
| true       | true     |

the `|| false` is thus redundant.

File: core/src/test/java/io/questdb/griffin/NullLiteralsTest.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin;
+import org.junit.Test;
```
Comment (marregui) [inline core/src/test/java/io/questdb/griffin/NullLiteralsTest.java:26]: needs a blank line between package declaration and first import.

File: core/src/test/java/io/questdb/griffin/InsertNullTest.java [hunk 5]
```diff
@@ -234,4 +234,18 @@ static String expectedNullInserts(String header, String nullValue, int count) {
         }
         return sb.toString();
     }
+
+    static String expectedNullInserts(String header, String type, String nullValue, int count, boolean eq) {
```
Comment (marregui) [inline core/src/test/java/io/questdb/griffin/InsertNullTest.java:238]: I would simply use one method called `expectedNullInserts` like so:

```
static String expectedNullInserts(String header, String nullValue, int count, boolean expectsOutput) {
        StringSink sb = Misc.getThreadLocalBuilder();
        sb.put(header);
        if (expectsOutput) {
            for (int i = 0; i < count; i++) {
                sb.put(nullValue).put("\n");
            }
        }
        return sb.toString();
    }

    private static boolean isNotNullable(String type) {
        return Chars.equalsLowerCaseAscii(type, "short") ||
                Chars.equalsLowerCaseAscii(type, "byte") ||
                Chars.equalsLowerCaseAscii(type, "boolean");
    }
```

then you can call it:

```
testInsertNullThenFilterIsNull: expectedNullInserts("value\n", type[1], NULL_INSERTS, !isNotNullable(type[0])),
testInsertNullThenFilterIsNotNull: expectedNullInserts("value\n", type[1], NULL_INSERTS, isNotNullable(type[0])),
```

Note, you will need to amend InsertNullGeoHashTest (add true at the end in L150)

**Review by marregui — CHANGES_REQUESTED** [review]
Overall very good, we are almost at the point of merging, how exciting!!!!! thank you for this Isabella.

**Comment by bluestreak01** [comment]
Hi Isabella, thank you for the PR! I had a look and it appears that the `Eq...` functions are not implemented to be efficient. When either of the args is `null constant` it renders the entire function `constant`. QuestDB SQL optimiser finds such constant functions and removes them from per-row execution.

The current implementations will be executing `where short = null` for each row. This should not be the case (it is `false` for every row).

I would suggest to move check for `null constant` to the factories. These factories should return `BooleanConstant.FALSE` if either of the args is `null constant`

**Comment by XinyiQiao** [comment]
Hi Vlad @bluestreak01 thank you so much for the feedback! Understood that the factories should return BooleanConstant.FALSE for short = null case. For the short != null case, I believe it should be true for every row. Could you advise on how to approach the != case? Where would be the best place to check for the negated flag?

**Comment by XinyiQiao** [comment]
> Hi Isabella, thank you for the PR! I had a look and it appears that the `Eq...` functions are not implemented to be efficient. When either of the args is `null constant` it renders the entire function `constant`. QuestDB SQL optimiser finds such constant functions and removes them from per-row execution.
> 
> The current implementations will be executing `where short = null` for each row. This should not be the case (it is `false` for every row).
> 
> I would suggest to move check for `null constant` to the factories. These factories should return `BooleanConstant.FALSE` if either of the args is `null constant`

Hi Vlad @bluestreak01 thank you so much for the feedback! I've moved the check for `null constant` to the factories. Please let me know if any additional issues need to be addressed.

**Review by marregui — APPROVED** [review]
LGTM well done @XinyiQiao

**Comment by marregui** [comment]
Hi @XinyiQiao , could you please update your fork and the branch, thank you in advance.

**Review by bluestreak01 — CHANGES_REQUESTED** [review]
performance improvement is required

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**